sympy/geometry/ellipse.py

```diff
@@ -1339,28 +1341,79 @@ class Circle(Ellipse):
     >>> c2.hradius, c2.vradius, c2.radius, c2.center
     (sqrt(2)/2, sqrt(2)/2, sqrt(2)/2, Point2D(1/2, 1/2))
 
+    >>> # a circle object is returned
+    >>> from sympy.abc import x, y, a, b
+    >>> Circle(x ** 2 + y ** 2 - 25)
+    Circle(Point2D(0, 0), 5)
+    >>> Circle(a ** 2 + b ** 2 - 25, x='a', y='b')
+    Circle(Point2D(0, 0), 5)
+
     """
 
     def __new__(cls, *args, **kwargs):
-        c, r = None, None
-        if len(args) == 3:
-            args = [Point(a, dim=2) for a in args]
-            if Point.is_collinear(*args):
-                raise GeometryError(
-                    "Cannot construct a circle from three collinear points")
-            from .polygon import Triangle
-            t = Triangle(*args)
-            c = t.circumcenter
-            r = t.circumradius
-        elif len(args) == 2:
-            # Assume (center, radius) pair
-            c = Point(args[0], dim=2)
-            r = sympify(args[1])
-
-        if not (c is None or r is None):
-            return GeometryEntity.__new__(cls, c, r, **kwargs)
-
-        raise GeometryError("Circle.__new__ received unknown arguments")
+
+        if len(args) == 1 and isinstance(args[0], Expr):
+            x = kwargs.get('x', 'x')
+            y = kwargs.get('y', 'y')
+            equation = args[0]
+
+            def find(x_, equation_):
+                free = equation_.free_symbols
+                xs = [i for i in free if (i.name if type(x_) is str else i) == x_]
+                if not xs:
+                    raise ValueError('could not find %s' % x_)
+                if len(xs) != 1:
+                    raise ValueError('ambiguous %s' % x_)
+                return xs[0]
+
+            x = find(x, equation)
+            y = find(y, equation)
+
+            try:
+                center_x = solve(diff(equation, x))[0]
+                center_y = solve(diff(equation, y))[0]
+            except IndexError:
+                raise GeometryError("The given equation is not that of a circle.")
+
+            center = Point(center_x, center_y)
+            constant_term = equation.subs([(x, 0), (y, 0)])
+            r_square = (center_x ** 2) + (center_y ** 2) - constant_term
+            new_equation = (x - center_x) ** 2 + (y - center_y) ** 2 - r_square
+
+            if r_square.is_nonnegative:
```

Comment:
perhaps allow for vanilla symbols: `if r_square.is_nonnegative is not False:` and then test with equation `x**2 + y**2 + c`. You don't know that `c` is positive so you can return a Circle assuming it isn't.

```python
>>> Circle((0,0),sqrt(-c))
Circle(Point2D(0, 0), sqrt(-c))
>>> Circle((0,0),sqrt(-c)).subs(c,-4)
Circle(Point2D(0, 0), 2)
```

Reply:
@smichr  Again, there will be no need to do that.
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> from sympy.abc import *
>>> Circle(x ** 2 + y ** 2 + 25)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/avi/sympy/sympy/geometry/ellipse.py", line 1390, in __new__
    raise GeometryError("The given equation of circle has an imaginary radius")
sympy.geometry.exceptions.GeometryError: The given equation of circle has an imaginary radius
>>> Circle(x ** 2 + y ** 2 - 25)
Circle(Point2D(0, 0), 5)
```
when `c` is negative, that exception gives out another `GeometryError`.
So negative `c` isn't a problem.

Reply:
```
>>> Circle(x**2+y**2+c)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\geometry\ellipse.py", line 1390, in __new__
    raise GeometryError("The given equation of circle has an imaginary radius")
sympy.geometry.exceptions.GeometryError: The given equation of circle has an imaginary radius
```

Reply:
@smichr  Oh! I didn't realise you meant `c` literally. Thanks for this input. Corrected now.
The out put is like this now:
```
avi@avi-Aspire-A515-51G:~/sympy$ python3
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> from sympy.abc import *
>>> Circle(x ** 2 + y ** 2 + c)
Circle(Point2D(0, 0), sqrt(-c))
```

Reply:
@smichr  Please take a look. Does this look like it can be merged now?

Reply:
@smichr Can you also restart the failing test? (I think it is failing because of some timeout error)